Feature: get all the ciphertexts from the LedgerState

IAvecilla opened the pull request:
<!-- Thank you for filing a PR! Help us understand by explaining your changes. Happy contributing! -->

## Motivation

Get all the ciphertexts directly from `LedgerState` to avoid iterating over all the transactions and commitments to get them.

## Test Plan

<!-- If you changed any code, please provide us with clear instructions on how you verified your changes work. -->

There is a new test created in `storage/src/state/tests.rs`
- `test_get_all_ciphertexts` - This test asserts that this function gets all the ciphertexts in the ledger without a specific order.

IAvecilla committed:
Add new function to get all ciphertexts from ledger state

IAvecilla committed:
Add test for get_ciphertexts function

IAvecilla committed:
Add documentation for each function

IAvecilla committed:
Refactor the test to compare sets of the ciphertexts to not assert the exact same order

IAvecilla committed:
Change return type of get ciphertexts method
- Now returns an iterator instead of a vector of ciphertexts.

ljedrz commented on storage/src/state/ledger.rs:
```diff
@@ -1567,6 +1577,14 @@ impl<N: Network> TransactionState<N> {
 
+    // Returns all the record ciphertexts.
+    fn get_ciphertexts(&self) -> Result<impl Iterator<Item = N::RecordCiphertext> + '_> {
+        Ok(self
+            .commitments
+            .keys()
+            .map(move |commitment| self.get_ciphertext(&commitment).unwrap()))
```
why does this function return a `Result` if we're calling `unwrap` on every item? we should probably return an `impl Iterator<Item = Result<N::RecordCiphertext>> + '_` instead

IAvecilla replied:
I agree with this. The change is already done on the last commit.

ljedrz commented on storage/src/state/tests.rs:
```diff
@@ -32,6 +32,14 @@ fn create_new_ledger<N: Network, S: Storage>() -> LedgerState<N> {
     LedgerState::open_writer_with_increment::<S, _>(temp_dir(), 1).expect("Failed to initialize ledger")
 }
 
+fn to_hash_set_test<N: Network>(ciphertexts: Vec<N::RecordCiphertext>) -> HashSet<N::RecordCiphertext> {
+    let mut set = HashSet::new();
+    for ciphertext in ciphertexts {
+        set.insert(ciphertext);
+    }
+    set
+}
```
I don't think this helper is needed, this is just `ciphertexts.into_iter().collect::<HashSet<_>>()`

IAvecilla replied:
You're right. I deleted this helper function and I changed the collect type on the last commit.

ljedrz commented on storage/src/state/tests.rs:
```diff
@@ -346,3 +354,34 @@ fn test_transaction_fees() {
+            .iter()
+            .flat_map(|block| block.commitments())
+            .map(|commitment| ledger.get_ciphertext(commitment).unwrap())
+            .collect::<Vec<_>>(),
```
```suggestion
            .collect::<HashSet<_>>(),
```
(which renders the call to `to_hash_set_test` redundant)

ljedrz commented on storage/src/state/tests.rs:
```diff
@@ -346,3 +354,34 @@ fn test_transaction_fees() {
+            .map(|commitment| ledger.get_ciphertext(commitment).unwrap())
+            .collect::<Vec<_>>(),
+    );
+    let ciphertexts_set = to_hash_set_test::<Testnet2>(ledger.get_ciphertexts().unwrap().collect());
```
```suggestion
    let ciphertexts_set = ledger.get_ciphertexts().unwrap().collect();
```

ljedrz commented on storage/src/state/tests.rs:
```diff
@@ -346,3 +354,34 @@ fn test_transaction_fees() {
+
+    // Mine the next block.
+    let (block, _) = ledger
+        .mine_next_block(address, true, &[], &terminator, rng)
```
instead of mining, we should load a single pre-existing block for a much faster testing experience, like in the `test_get_blocks` test in `snarkos-rpc`.

ljedrz replied:
note to self: we should make sure we're not mining in _any_ test unless absolutely necessary.

IAvecilla replied:
Great! I was mining a block to ensure that the function was getting all the actual ciphertexts but I think the test works well with only the genesis block. I deleted that part of the test on the last commit.

ljedrz replied:
BTW, I already proposed the aforementioned change in https://github.com/AleoHQ/snarkOS/pull/1708/commits/c45463aa98f8c6ed64962647c3183f36f172893c.

ljedrz commented:
Left some comments; also, these changes should be squashed into 1-2 commits.

IAvecilla committed:
Make suggested changes for get ciphertexts function (#2)

* Change return type of get ciphertexts function
- The return type change to an iterator of ciphertexts results to avoid unwraps

* Delete test helper to create a set of ciphertext

* Fix test to use new return type of get ciphertexts function

* Delete mining block process from the get ciphertexts test
